src/agreement/mod.rs

```diff
@@ -489,15 +493,21 @@ impl<NodeUid: Clone + Debug + Ord> Agreement<NodeUid> {
         sender_id: &NodeUid,
         msg: CommonCoinMessage,
     ) -> AgreementResult<FaultLog<NodeUid>> {
-        let mut fault_log = self.common_coin.handle_message(sender_id, msg)?;
+        let coin_step = self.common_coin.handle_message(sender_id, msg)?;
         self.extend_common_coin();
+        self.on_coin_step(coin_step)
+    }
 
-        if let Some(coin) = self.common_coin.next_output() {
+    fn on_coin_step(
+        &mut self,
+        coin_step: CommonCoinStep<NodeUid>,
+    ) -> AgreementResult<FaultLog<NodeUid>> {
+        let mut fault_log = FaultLog::new();
+        fault_log.extend(coin_step.fault_log);
```

Comment:
You could clone out the `FaultLog` from `coin_step` and save yourself a line:

```rust
let mut fault_log = coin_step.fault_log.clone();
```

Reply:
Maybe we could `drain` the log instead of cloning it? Might go in a separate issue.